POI Service Supported (#877)
smartdevicelink/sdl_evolution: proposals/NNNN-POI-Service-Supported.md

# POI Service Supported

* Proposal: [SDL-NNNN](NNNN-POI-Service-Supported.md)
* Author: [Shohei Kawano](https://github.com/Shohei-Kawano)
* Status: Awaiting review
* Impacted Platforms: [ Core / iOS / Java Suite / RPC ]

## Introduction

This porposal provide the function of sharing POI data between SDL App by `AppService`.  

## Motivation

Recently, IVI has moved from AVN (Audio & Visual Navigation) to DA (Display Audio).  
Therefore, when there is no Native Navigation, it is necessary to share POIs between SDL apps.  
The proposal of [SDL-0241 Commerce App Service](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0241-Commerce-App-Service.md) included a function to provide search data to other apps,   
but there was no specific use case and the function was excessive. In this proposal,   
only the minimum necessary POI exchange is provided for the App service.  
In the future, if someone want to increase the data exchanged between apps,   

> it is assumed that proposals and extensions will be made each time.  

Author: commit b0952808bf06f8ffc523fc3f647156c6de9e9e3f (HEAD -> master, origin/master, origin/HEAD)
Author: 河野 祥平 <103082@ad.nexty-ele.com>
Date:   Thu Dec 5 17:48:07 2019 +0900

    remove the extra spacing here, and make this section one paragraph

> it is assumed that proposals and extensions will be made each time.  

Reviewer: I've added a new comment to address this concern.

***
[Assumed use case]
1. Launch POI app  
2. Search POI  
3. Shown POI list  
4. Choise from POI list  
5. Set the choised POI data to struct of `PoiServiceData->LoacationDetails`  
6. Shown popup message that is able to choise following purpose  
 (a) Set as destination ( Share the POI to SDL Navigation App )  
 (b) Call to the phone number included in the POI  

When (a) is selected  
 7. If AppHMIType is "NAVIGATION", or if it is already LIMITED, go to (ii) -10.  
 8. When there are multiple items including Native Navi, Popup is displayed.  
 9. Choise the Native Navigation or the app.  
  
  (i) When Native Navigation is selected.  
   10. Set POI to RPC `SendLocation->LocationDetails`.  
   11. Send RPC `SendLication`  
   12. Shown Native Naigation screen and find route.  
  
  (ii) When the SDL navigation app is selected or there is an app with `HMIStatus` already LIMITED.  
   10. Activate or change `HMIStatus` to FULL the target SDL Navigation App  
   11. Subscribe to App Service from Navi App  
   12. Navi App get `PoiServiceData->LocationDetails`.  

>    13. Find route with Navi App.  

Reviewer: ```suggestion
   13. Find route with Navigation App.  
```

  
When (b) is selected  
 7. Set `PoiServiceData->LocationDetails->phoneNumber` to RPC `DialNumber->number`.  
 8. Send `DialNumber`.  
 9. Start phone call.  
***

> Since this proposal is a proposal for sharing POI data, it is assumed that Navi App has already been started.  

Reviewer: ```suggestion
Since this proposal is a proposal for sharing POI data, it is assumed that Navigation App has already been started.  
```

> So, navi selection and specified Nav iApp launch methods are not defined.  

Reviewer: ```suggestion
So, navigation selection and specified Navigation App launch methods are not defined.  
```

## Proposed solution

> Add new App Serivice Type for using this POI sharing method.  

Reviewer: ```suggestion
Add new App Service Type for using this POI sharing method.  
```

  
Add POI element to AppServiceType enum of both Mobile API and HMI API.

Mobile API:
```xml
    <enum name="AppServiceType" platform="documentation" since="5.1">
         <element name = "MEDIA"/>
         <element name = "WEATHER"/>
         <element name = "NAVIGATION"/>
+        <element name = "POI"/>
    </enum>
```

HMI API:
```xml
    <enum name="AppServiceType">
         <element name="MEDIA"/>
         <element name="WEATHER"/>
         <element name="NAVIGATION"/>
+        <element name="POI"/>
    </enum>

```
The POI service is defined as a service that can provide searched information.  

> As the information to be provided, the name, coordinates, address, outline of the place, and telephone number of the searched place are assumed.  

Reviewer: ```suggestion
As the information is to be provided, the name, coordinates, address, outline of the place, and telephone number of the searched place are assumed.  
```

This information matches the parameter of structure `Location Details` defined in the navigation.  

> Mobile API(struct):

Reviewer: ```suggestion
Mobile API (struct):
```

	
	<param name="serviceType" type="String" mandatory="true">
	<description>The type of service that is to be offered by this app. See AppServiceType for known enum equivalent types. Parameter is a string to allow for new service types to be used by apps on older versions of SDL Core.</description>
	</param>
	<param name="serviceID" type="String" mandatory="true"/>
	
	<param name="mediaServiceData" type="MediaServiceData" mandatory="false"/>
	<param name="weatherServiceData" type="WeatherServiceData" mandatory="false"/>
	<param name="navigationServiceData" type="NavigationServiceData" mandatory="false"/>
+	<param name="poiServiceData" type="PoiServiceData" mandatory="false"/>
</struct>	
	
+<struct name="PoiServiceManifest" since="5.1">	
+	<description> This manifest contains all the information necessary for the service to be published, activated, and consumers able to interact with it </description>
+	
+	<param name="poiInfoDataSupported" type="Boolean" mandatory="false">

> +	<description> Informs the subscriber if this service can actually availble. </description>

Reviewer: ```suggestion
+	<description> Informs the subscriber if this service is actually available. </description>
```

+	</param>
+</param>	
	
+<struct name="PoiServiceData" since="5.1">	

> +	<description> This data is related to what a POI(Points of Interest) service would provide. </description>

Reviewer: ```suggestion
+	<description> This data is related to what a POI (Point of Interest) service would provide. </description>
```

+	
+	<param name="poiInfoData" type="LocationDetails" mandatory="false">

> +	<description> Information data of searched by POI. See LocationDetails </description>

Reviewer: ```suggestion
+	<description> Information data of searched POI. See LocationDetails </description>
```

+	</param>
+</param>	
```

> HMI API(struct):

Reviewer: ```suggestion
HMI API (struct):
```

	
	<param name="serviceType" type="String" mandatory="true">
	<description>The type of service that is to be offered by this app. See AppServiceType for known enum equivalent types. Parameter is a string to allow for new service types to be used by apps on older versions of SDL Core.</description>
	</param>
	<param name="serviceID" type="String" mandatory="true"/>
	
	<param name="mediaServiceData" type="Common.MediaServiceData" mandatory="false"/>
	<param name="weatherServiceData" type="Common.WeatherServiceData" mandatory="false"/>
	<param name="navigationServiceData" type="Common.NavigationServiceData" mandatory="false"/>
+	<param name="poiServiceData" type="Common.PoiServiceData" mandatory="false"/>
</struct>	
	
+<struct name="PoiServiceManifest">	
+	<description> This manifest contains all the information necessary for the service to be published, activated, and consumers able to interact with it </description>
+	
+	<param name="poiInfoDataSupported" type="Boolean" mandatory="false">

> +	<description> Informs the subscriber if this service can actually availble. </description>

Reviewer: ```suggestion
+	<description> Informs the subscriber if this service is actually available. </description>
```

+	</param>
+</param>	
	
+<struct name="PoiServiceData">	

> +	<description> This data is related to what a POI(Points of Interest) service would provide. </description>

Reviewer: ```suggestion
+	<description> This data is related to what a POI (Point of Interest) service would provide. </description>
```

+	
+	<param name="poiInfoData" type="LocationDetails" mandatory="false">

> +	<description> Information data of searched by POI. See LocationDetails </description>

Reviewer: ```suggestion
+	<description> Information data of searched POI. See LocationDetails </description>
```

+	</param>
+</param>	
```

## Potential downsides

Nothing.  

## Impact on existing code

> As new application services are added, Core, iOS, Java Suite, and RPC need to be modified.  

Reviewer: Can you provide more information about the modifications required?

Author: commit adce0350ba221dd84944c6a5a135e5ba11e12477 (HEAD -> master, origin/master, origin/HEAD)
Author: 河野 祥平 <103082@ad.nexty-ele.com>
Date:   Fri Dec 6 08:22:01 2019 +0900

    Modify Impact on existing code. and modify Alternaties considerd

* Proposal: [SDL-NNNN](NNNN-POI-Service-Supported.md)
* Author: [Shohei Kawano](https://github.com/Shohei-Kawano)
* Status: Awaiting review
* Impacted Platforms: [ Core / RPC ]

## Introduction

This proposal provides the function of sharing POI data between SDL Apps by `AppService`.  

## Motivation

Recently, IVI has moved from AVN (Audio & Visual Navigation) to DA (Display Audio).  
Therefore, when there is no Native Navigation, it is necessary to share POIs between SDL apps.  
The proposal of [SDL-0241 Commerce App Service](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0241-Commerce-App-Service.md) included a function to provide search data to other apps, but there was no specific use case and the function was excessive.  

> In this proposal, only the minimum necessary POI exchange is provided for the App service.  

Reviewer: Please change this line to the following:
>In this proposal, only the minimum necessary POI exchange is provided for the App service. In the future, if someone wants to increase the data exchanged between apps, it is assumed that proposals and extensions will be made each time.  

You can then remove lines 19 and 20.  This will help the section to be formatted in a more readable way.

Author: modified!

commit 992b7e3435acf4ed2391f6c059b8cc8d5730b62a (HEAD -> master, origin/master, origin/HEAD)
Author: 河野 祥平 <103082@ad.nexty-ele.com>
Date:   Wed Dec 11 13:34:40 2019 +0900

    Remove unnecessary break
